src/include/daos_fs_sys.h

```diff
@@ -0,0 +1,456 @@
+
+/**
+ * Retrieve Symlink value of path if it's a symlink. If the buffer size passed
+ * in is not large enough, we copy up to size of the buffer, and update the
+ * size to actual value size. The size returned includes the null terminator.
+ *
+ * \param[in]	dfs_sys Pointer to the mounted file system.
+ * \param[in]	path	Link path of object.
+ * \param[in,out]
+ *		buf	[in]: Allocated buffer for value.
+ *			[out]: Symlink value.
+ * \param[in,out]
+ *		size	[in]: Size of buffer passed in. [out]: Actual size of
+ *			      value. On error, this is set to -1.
+ *
+ * \return		0 on success, errno code on failure.
+ */
+int
+dfs_sys_readlink(dfs_sys_t *dfs_sys, const char *path, char *buf,
+		 daos_size_t *size);
+
+/**
+ * Create a symlink.
+ *
+ * \param[in]	dfs_sys Pointer to the mounted file system.
+ * \param[in]	target	Symlink value.
+ * \param[in]	path	Path to the new symlink.
+ *
+ * \return		0 on success, errno code on failure.
+ */
+int
+dfs_sys_symlink(dfs_sys_t *dfs_sys, const char *target, const char *path);
+
+/**
+ * Create/Open a directory, file, or Symlink.
+ * The object must be released with dfs_sys_close().
+ *
+ * \param[in]	dfs_sys Pointer to the mounted file system.
+ * \param[in]	path	Link path of the object to create/open.
+ * \param[in]	mode	mode_t (permissions + type).
+ * \param[in]	flags	Access flags (handles: O_RDONLY, O_RDWR, O_EXCL,
+ *			O_CREAT, O_TRUNC).
+ * \param[in]	cid	DAOS object class id (pass 0 for default MAX_RW).
+ *			Valid on create only; ignored otherwise.
+ * \param[in]	chunk_size
+ *			Chunk size of the array object to be created.
+ *			(pass 0 for default 1 MiB chunk size).
+ *			Valid on file create only; ignored otherwise.
+ * \param[in]	value	Symlink value (NULL if not syml).
+ * \param[in]	obj	Pointer to object opened.
+ *
+ * \return		0 on success, errno code on failure.
+ */
+int
+dfs_sys_open(dfs_sys_t *dfs_sys, const char *path, mode_t mode, int flags,
+	     daos_oclass_id_t cid, daos_size_t chunk_size,
+	     const char *value, dfs_obj_t **obj);
+
+/**
+ * Close/release open object.
+ *
+ * \param[in]	obj	Object to release.
+ *
+ * \return		0 on success, errno code on failure.
+ */
+int
+dfs_sys_close(dfs_obj_t *obj);
+
+/**
+ * Read data from the file object, and return actual data read.
+ *
+ * \param[in]	dfs_sys Pointer to the mounted file system.
+ * \param[in]	obj	Opened file object.
+ * \param[in,out]
+ *		buf	[in]: Allocated buffer for data.
+ *			[out]: Actual data read.
+ * \param[in]	off	Offset into the file to read from.
+ * \param[in,out]
+ *		size	[in]: Size of buffer passed in. [out]: Actual size of
+ *			      data read. On error, this is set to -1.
+ * \param[in]	ev	Completion event, it is optional and can be NULL.
+ *			Function will run in blocking mode if \a ev is NULL.
+ *
+ * \return		0 on success, errno code on failure.
+ */
+int
+dfs_sys_read(dfs_sys_t *dfs_sys, dfs_obj_t *obj, void *buf, daos_off_t off,
```

Comment:
would be good to add the readx and writex APIs too.

Reply:
Added. These are really just "direct" wrappers to the dfs versions.

Reply:
For read and write, and maybe ostat would it be worth having a "dfs_sys_get_dfs_obj()" function that would return something that allowed you to call the dfs api directly?

Reply:
@ashleypittman Definitely could be useful. Right now, the dfs_sys struct exposes the underlying dfs instance for exactly this purpose. If we allow calling the dfs api directly, then we don't have to force users to pick EITHER the dfs OR the dfs_sys api - they could use both here. Generally, if a user already has an open handle to a file/dir, then they could just call the dfs versions directly. So I think a dfs_sys_get_dfs_obj function would be good.

@mchaarawi What do you think?

Reply:
what is the use case for this though?
just to be clear, im fine to add this API. However it is cleaner to add the read/write(x) wrappers and not expect users to need to call the sys->base DFS call to be able to call the IO calls.

Reply:
The dfs functions - in some cases - can be more efficient than the dfs_sys functions if you already have the open object handle. For example, dfs_ostat, dfs_osetattr.
I see your point about it being cleaner, though. Would it makes sense to add the wrappers AND also a sort of dfs_sys_get_dfs_obj?
Something we would have to keep in mind is that when new functions are added to the DFS API, we would probably want to make sure they are accessible through dfs_sys.

Reply:
I was thinking it would save the effort of writing/maintaining the wrappers if they're just pass-through functions.

Reply:
yea i agree with that i guess. I'll leave it to Dalton if he wants to keep the io functions then.
should just comment somewhere clearly that to do IO with this layer we need this sys->base conversion.

Reply:
I'll add a dfs_sys_get_dfs_obj() function, and NOT keep the pass-through functions for now, since we could always add them later.
And I'll document this as well.